controller/volume_controller.go

```diff
@@ -771,33 +782,88 @@ func (vc *VolumeController) cleanupFailedToScheduledReplicas(v *longhorn.Volume,
 
 func (vc *VolumeController) cleanupExtraHealthyReplicas(v *longhorn.Volume, e *longhorn.Engine, rs map[string]*longhorn.Replica) (err error) {
 	healthyCount := vc.getHealthyReplicaCount(rs)
+	if healthyCount <= v.Spec.NumberOfReplicas {
+		return nil
+	}
 
-	if healthyCount > v.Spec.NumberOfReplicas {
-		for _, r := range rs {
-			// Clean up eviction requested replica
-			if r.Status.EvictionRequested {
-				if err := vc.deleteReplica(r, rs); err != nil {
-					vc.eventRecorder.Eventf(v, v1.EventTypeWarning,
-						EventReasonFailedEviction,
-						"volume %v failed to evict replica %v",
-						v.Name, r.Name)
-					return err
-				}
-				logrus.Debugf("Evicted replica %v", r.Name)
-				return nil
+	var cleaned bool
+	if cleaned, err = vc.cleanupEvictionRequestedReplicas(v, rs); err != nil || cleaned {
+		return err
+	}
+
+	if cleaned, err = vc.cleanupAutoBalancedReplicas(v, rs); err != nil || cleaned {
+		return err
+	}
+
+	if cleaned, err = vc.cleanupDataLocalityReplicas(v, e, rs); err != nil || cleaned {
+		return err
+	}
+
+	return nil
+}
+
+func (vc *VolumeController) cleanupEvictionRequestedReplicas(v *longhorn.Volume, rs map[string]*longhorn.Replica) (bool, error) {
+	for _, r := range rs {
+		if r.Status.EvictionRequested {
+			if err := vc.deleteReplica(r, rs); err != nil {
+				vc.eventRecorder.Eventf(v, v1.EventTypeWarning,
+					EventReasonFailedEviction,
+					"volume %v failed to evict replica %v",
+					v.Name, r.Name)
+				return false, err
 			}
+			logrus.Debugf("Evicted replica %v", r.Name)
+			return true, nil
 		}
 	}
+	return false, nil
+}
 
-	// Clean up extra healthy replica with data-locality finished
-	healthyCount = vc.getHealthyReplicaCount(rs)
+func (vc *VolumeController) cleanupAutoBalancedReplicas(v *longhorn.Volume, rs map[string]*longhorn.Replica) (bool, error) {
+	log := getLoggerForVolume(vc.logger, v).WithField("replicaAutoBalanceType", "delete")
+
+	setting, err := vc.getAutoBalancedReplicasSetting(v)
+	if err != nil {
+		return false, err
+	}
+	if setting == types.ReplicaAutoBalanceDisabled {
+		return false, nil
+	}
+
+	var rNames []string
+	if setting == types.ReplicaAutoBalanceBestEffort {
+		_, rNames = vc.replicaAutoBalanceBestEffort(v, rs, vc.replicaAutoBalanceZoneCount)
+		if len(rNames) == 0 {
+			_, rNames = vc.replicaAutoBalanceBestEffort(v, rs, vc.replicaAutoBalanceNodeCount)
+		}
+	}
+
+	if len(rNames) == 0 {
+		rNames, err = vc.getPreferredReplicaCandidatesForDeletion(rs)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	// Randomly delete extra non-local healthy replicas in the preferred candidate list rNames
+	// Sometime cleanupExtraHealthyReplicas() is called more than once with the same input (v,e,rs).
+	// To make the deleting operation idempotent and prevent deleting more replica than needed,
+	// we always delete the replica with the smallest name.
+	sort.Strings(rNames)
+	rName := rs[rNames[0]]
+	if err := vc.deleteReplica(rName, rs); err != nil {
```

Comment:
Deleting 1 replica is intended. 

We should consider to have more self-explanatory methods instead of implicit behavior inside like this. Not your problem, but we can enhance later.